### PR TITLE
Small bits of completion cleanup.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Review one commit at a time.

A core part of this change is updating how we compute preselection for symbol completion providers.  Previously we computed completion items in two passes.  First, computing the non-preselected-items, then the preselected-items.   This was complex, and forced us to pass a lot of data around to keep track of this approach.  The new way of doing things is just a simple approach where the subclass just returns teh set of symbols, along with a value per symbol if it should be preselected or not.  This ends up being much simpler and clearer overall. 

Review with whitespace changes off.

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs [hunk 1]
```diff
@@ -30,31 +29,75 @@ namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
     [Shared]
     internal partial class SymbolCompletionProvider : AbstractRecommendationServiceBasedCompletionProvider
     {
+        private static readonly Dictionary<(bool importDirective, bool preselect, bool tupleLiteral), CompletionItemRules> s_cachedRules = new();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:32]: this is an unchanged move-up of code from below

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs [hunk 2]
```diff
@@ -30,31 +29,75 @@ namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
...
+            if (context != null)
             {
-                // Avoid preselection & hard selection when triggered via insertion in an argument list.
-                // If an item is hard selected, then a user trying to type MethodCall() will get
-                // MethodCall(someVariable) instead. We need only soft selected items to prevent this.
-                if (completionContext.Trigger.Kind == CompletionTriggerKind.Insertion &&
-                    position > 0 &&
-                    await IsTriggerInArgumentListAsync(document, position - 1, CancellationToken.None).ConfigureAwait(false) == true)
+                var document = context.Document;
+                var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:89]: we weren't passing along cancellation :-/

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs [hunk 3]
```diff
@@ -97,17 +136,6 @@ internal override async Task<bool> IsSyntacticTriggerCharacterAsync(Document doc
 
         internal override ImmutableHashSet<char> TriggerCharacters { get; } = CompletionUtilities.CommonTriggerCharactersWithArgumentList;
 
-        protected override async Task<bool> IsSemanticTriggerCharacterAsync(Document document, int characterPosition, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:100]: this logic moved up to base class.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs [hunk 4]
```diff
@@ -132,23 +160,25 @@ private bool ShouldTriggerInArgumentLists(Workspace workspace, OptionSet options
...
-                token = token.GetPreviousToken();
-            }
+                return true;
+
+            // if we're right after the first dot in .. then that's considered completion on dot.
+            if (token.Kind() == SyntaxKind.DotDotToken && token.SpanStart == characterPosition)
+                return true;
 
-            return token.Kind() != SyntaxKind.NumericLiteralToken;
+            return false;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:181]: the subclasses only have to answer questions about the dot token.  shared logic for vb/c# moved up.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs [hunk 5]
```diff
@@ -196,79 +226,21 @@ protected override async Task<SyntaxContext> CreateContextAsync(Document documen
         protected override (string displayText, string suffix, string insertionText) GetDisplayAndSuffixAndInsertionText(ISymbol symbol, SyntaxContext context)
             => CompletionUtilities.GetDisplayAndSuffixAndInsertionText(symbol, context);
 
-        protected override CompletionItemRules GetCompletionItemRules(List<ISymbol> symbols, SyntaxContext context, bool preselect)
+        protected override CompletionItemRules GetCompletionItemRules(ImmutableArray<(ISymbol symbol, bool preselect)> symbols, SyntaxContext context)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:229]: several APIs moved to immutable collections.

File: src/Features/Core/Portable/Completion/CommonCompletionProvider.cs [hunk 6]
```diff
@@ -17,6 +17,8 @@ namespace Microsoft.CodeAnalysis.Completion
 {
     internal abstract class CommonCompletionProvider : CompletionProvider
     {
+        private static readonly CompletionItemRules s_suggestionItemRules = CompletionItemRules.Create(enterKeyRule: EnterKeyRule.Never);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CommonCompletionProvider.cs:20]: i tried to move statics up to the top of classes.

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectCreationCompletionProvider.cs [hunk 7]
```diff
@@ -116,22 +88,19 @@ protected override Task<ImmutableArray<ISymbol>> GetSymbolsAsync(SyntaxContext c
...
             if (!type.IsEditorBrowsable(options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language), context.SemanticModel.Compilation))
-            {
-                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
-            }
+                return SpecializedTasks.EmptyImmutableArray<(ISymbol symbol, bool preselect)>();
 
-            return Task.FromResult(ImmutableArray.Create((ISymbol)type));
+            // In the case of array creation, we don't offer a preselected/hard-selected item because
+            // the user may want an implicitly-typed array creation
+            return Task.FromResult(ImmutableArray.Create(((ISymbol)type, preselect: !isArray)));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectCreationCompletionProvider.cs:103]: this is an example of code we would run twice.  once for non-preselect, and once for preselect.  Now we just run things once, but determien if we should preselect or not just once.

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs [hunk 8]
```diff
@@ -108,12 +107,6 @@ private static bool ShouldSoftSelectInArgumentList(CompletionContext completionC
         private static bool IsArgumentListTriggerCharacter(char character)
             => character == ' ' || character == '(' || character == '[';
 
-        protected abstract CompletionItemRules GetCompletionItemRules(List<ISymbol> symbols, SyntaxContext context, bool preselect);
-
-        protected abstract CompletionItemSelectionBehavior PreselectedItemSelectionBehavior { get; }
-
-        protected abstract bool IsInstrinsic(ISymbol symbol);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs:115]: moved abstracts up.

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs [hunk 9]
```diff
@@ -281,16 +278,8 @@ public override async Task ProvideCompletionsAsync(CompletionContext completionC
...
+                    var regularItems = await GetItemsWorkerAsync(completionContext, syntaxContext, document, position, options, telemetryCounter, cancellationToken).ConfigureAwait(false);
                     completionContext.AddItems(regularItems);
-
-                    // We may or may not want to provide preselected items based on context. For example, C# avoids provide providing preselected items when
-                    // triggered via text insertion in an argument list.
-                    if (await ShouldProvidePreselectedItemsAsync(completionContext, syntaxContext, document, position, options).ConfigureAwait(false))
-                    {
-                        var preselectedItems = await GetItemsWorkerAsync(completionContext, syntaxContext, document, position, options, preselect: true, telemetryCounter, cancellationToken).ConfigureAwait(false);
-                        completionContext.AddItems(preselectedItems);
-                    }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:293]: this is hte core change in the PR.  Instead of a two-pass approach, we just have GetItemsWorkerAsync return all the data we need.

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/SymbolCompletionProvider.vb [hunk 10]
```diff
@@ -21,54 +21,72 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
...
+
+        Shared Sub New()
+            For importDirective = 0 To 1
+                For preselect = 0 To 1
+                    For tuple = 0 To 1
+                        Dim context = (importDirective:=importDirective = 1, preselect:=preselect = 1, tuple:=tuple = 1)
+                        s_cachedRules(context) = MakeRule(context)
+                    Next
+                Next
+            Next
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/SymbolCompletionProvider.vb:35]: this was just code from below that moved up.

**Comment by genlu** [comment]
>                 var resolvableMembers = overridableMembers.Where(m => CanResolveSymbolKey(m, semanticModel.Compilation));

Shouldn't the fix be using the result of `CanResolveSymbolKey` instead of remove this step completely?

---
Refers to: src/Features/Core/Portable/Completion/Providers/AbstractOverrideCompletionProvider.ItemGetter.cs:92 in f7a5384. [](commit_id = f7a53840fc3c01e61aad1d0ba23369ef75437d12, deletion_comment = True)

**Comment by CyrusNajmabadi** [comment]
> Shouldn't the fix be using the result of CanResolveSymbolKey instead of remove this step completely?

I thought about that. However, that code seems to be completely unused and untested. So.... Keeping it would just be adding new code that wasn't every needed before... So... :-)

**Review by genlu — APPROVED** [review]
:shipit:

File: src/Features/Core/Portable/Completion/CommonCompletionProvider.cs [hunk 11]
```diff
@@ -93,14 +95,12 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
...
-        private static readonly CompletionItemRules s_suggestionItemRules = CompletionItemRules.Create(enterKeyRule: EnterKeyRule.Never);
-
-        protected static CompletionItem CreateSuggestionModeItem(string displayText, string description)
+        protected static CompletionItem CreateSuggestionModeItem(string? displayText, string? description)
         {
             return CommonCompletionItem.Create(
                 displayText: displayText ?? string.Empty,
                 displayTextSuffix: "",
-                description: description != null ? description.ToSymbolDisplayParts() : default,
+                description: description == null ? default : description.ToSymbolDisplayParts(),
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/CommonCompletionProvider.cs:103]:
```suggestion
                description: description?.ToSymbolDisplayParts() ?? default,
```

File: src/Features/Core/Portable/Completion/CommonCompletionProvider.cs [hunk 12]
```diff
@@ -29,7 +31,7 @@ public override bool ShouldTriggerCompletion(SourceText text, int position, Comp
             }
         }
 
-        internal virtual bool IsInsertionTrigger(SourceText text, int insertedCharacterPosition, OptionSet options)
+        public virtual bool IsInsertionTrigger(SourceText text, int insertedCharacterPosition, OptionSet options)
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/CommonCompletionProvider.cs:34]: ❔ Why make this public?

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 13]
```diff
@@ -62,31 +62,28 @@ public static Document ReplaceNode<TNode>(this Document document, SyntaxNode roo
...
-            var itemsForCurrentContext = await getItemsWorker(document, cancellationToken).ConfigureAwait(false) ?? SpecializedCollections.EmptyEnumerable<T>();
-            if (!linkedDocumentIds.Any())
-            {
-                return itemsForCurrentContext;
-            }
+            using var _ = ArrayBuilder<Task<ImmutableArray<T>>>.GetInstance(out var tasks);
+            tasks.Add(Task.Run(() => getItemsWorker(document), cancellationToken));
+
+            foreach (var linkedDocumentId in document.GetLinkedDocumentIds())
+                tasks.Add(Task.Run(() => getItemsWorker(document.Project.Solution.GetRequiredDocument(linkedDocumentId)), cancellationToken));
```
Comment (sharwell) [inline src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs:75]: ❗ Parallelism strategy change needs to be in a standalone PR

**Review by sharwell — CHANGES_REQUESTED** [review]
Parallelism changes shouldn't be merged with refactoring PRs

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SnippetCompletionProvider.cs [hunk 14]
```diff
@@ -133,25 +132,24 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
...
                 {
-                    return await GetSnippetCompletionItemsAsync(workspace, semanticModel, isPreProcessorContext: false,
+                    return await GetSnippetCompletionItemsAsync(
+                        document.Project.Solution.Workspace, semanticModel, isPreProcessorContext: false,
                         isTupleContext: isPossibleTupleContext, cancellationToken: cancellationToken).ConfigureAwait(false);
                 }
             }
 
-            return SpecializedCollections.EmptyEnumerable<CompletionItem>();
+            return default;
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/SnippetCompletionProvider.cs:141]: ❔ Why default instead of empty?

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs (hunk 13 above)
Comment (genlu) [reply src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs:75]: I think this might be less of an issue compare to the similar change @CyrusNajmabadi made in keyword completion, since the work here is mostly likely to be completed asynchronously and the number of linked document is much smaller.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:88]: We shouldn't need `sourceText` at all. The only thing it's used for is getting `IExperimentationService`, and we can get that directly from `document`.

File: src/Features/Core/Portable/Completion/CommonCompletionProvider.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/CommonCompletionProvider.cs:34]: because it's in an internal type.  that's the general pattern we use.

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs:75]: i don't mind moving into another pr.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SnippetCompletionProvider.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/SnippetCompletionProvider.cs:141]: Def should be empty instead :)

File: src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:88]: yup.  got rid of the need for this.

File: src/Features/Core/Portable/Completion/CommonCompletionProvider.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/CommonCompletionProvider.cs:103]: thanks! i do prefer the ternary form though.  i really wish tehre was a `?.` form that gave you a `default(T)` not a `T?`.

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: have pulled out the parallelism bits.
